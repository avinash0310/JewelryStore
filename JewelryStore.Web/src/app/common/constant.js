export const REQUEST_AUTHENTICATION = 'REQUEST_AUTHENTICATION';
export const RECEIVE_AUTHENTICATION = 'RECEIVE_AUTHENTICATION';
export const RECEIVE_AUTHENTICATION_ERROR = 'RECEIVE_AUTHENTICATION_ERROR';
export const CLEAR_AUTHENTICATION_DATA = 'CLEAR_AUTHENTICATION_DATA';
export const API_BASE_ADDRESS = 'https://localhost:44348';
export const LOGIN_API_URL = API_BASE_ADDRESS + '/api/Authenticate/Login';
export const JEWELRY_STORE_PATH = '/jewelrystore';
export const LOGIN_PAGE_PATH = '/login';
export const DEFAULT_PAGE_PATH = '/';
export const POST = 'POST';
export const TOKEN = 'token';
export const ENDPOINT = 'endpoint';
export const PRIVILEGED = 'Privileged';
export const MIN_LENGTH = 5;
export const MAX_LENGTH = 12;
export const DEFAULT_DISCOUNT = 2;
export const CALCULATE_API_URL = API_BASE_ADDRESS + '/api/JewelryStore/Calculate';
export const REQUEST_JEWELRY_STORE_CALCULATION = 'REQUEST_JEWELRY_STORE_CALCULATION';
export const RECEIVE_JEWELRY_STORE_CALCULATION = 'RECEIVE_JEWELRY_STORE_CALCULATION';
export const RECEIVE_JEWELRY_STORE_CALCULATION_ERROR = 'RECEIVE_JEWELRY_STORE_CALCULATION_ERROR';
export const GOLD_PRICE_VALIDATION_MSG = 'Please provide valid Gold Price.';
export const GOLD_WEIGHT_VALIDATION_MSG = 'Please provide valid Gold Weight.';
export const DISCOUNT_VALIDATION_MSG = 'Please provide valid Discount Rate.';
export const INVALID_CREDENTIAL_VALIDATION_MSG = 'Please provide valid credential.';
export const PLEASE_CALCULATE_VALIDATION_MSG = 'Please calculate before print.';
export const INVALID_USER_NAME_VALIDATION_MSG = 'The field username must be a string with a minimum length of 5 and a maximum length of 12.';
export const INVALID_PASSWORD_VALIDATION_MSG = 'The field password must be a string with a minimum length of 5 and a maximum length of 10.';
export const FAILED_MESSAGE = 'Some error occurred please try again.';