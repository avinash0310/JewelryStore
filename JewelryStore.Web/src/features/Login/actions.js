import {
    REQUEST_AUTHENTICATION,
    RECEIVE_AUTHENTICATION,
    RECEIVE_AUTHENTICATION_ERROR,
    CLEAR_AUTHENTICATION_DATA,
    LOGIN_API_URL, POST
} from '../../app/common/constant.js';


export const requestLogin = (loginModel) => {
    return {
        type: REQUEST_AUTHENTICATION,
        endpoint: LOGIN_API_URL,
        verb: POST,
        headers: { 'Content-type': 'application/json' },
        payload: JSON.stringify(loginModel),
        onSuccess: receiveLoginResponse,
        onFail: receiveLoginError
    };
};

export const receiveLoginResponse = (payload) => {
    return {
        type: RECEIVE_AUTHENTICATION,
        payload: payload
    };
};

export const receiveLoginError = (error) => {
    return {
        type: RECEIVE_AUTHENTICATION_ERROR,
        error: error
    };
};

export const clearLoginData = () => {
    return {
        type: CLEAR_AUTHENTICATION_DATA
    };
};