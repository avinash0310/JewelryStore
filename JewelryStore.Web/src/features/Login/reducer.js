import {
    REQUEST_AUTHENTICATION,
    RECEIVE_AUTHENTICATION,
    RECEIVE_AUTHENTICATION_ERROR,
    CLEAR_AUTHENTICATION_DATA
} from '../../app/common/constant.js';


const INITIAL_STATE = {
    payload: null,
    isFailed: false,
    isLoading: false,
    error: null
};

const loginReducer = (state = INITIAL_STATE, action) => {
    switch (action.type) {
        case REQUEST_AUTHENTICATION:
            return {
                ...state,
                payload: null,
                isFailed: false,
                isLoading: true,
                error: null
            };

        case RECEIVE_AUTHENTICATION:
            return {
                ...state,
                payload: action.payload,
                isFailed: false,
                isLoading: false,
                error: null
            };

        case RECEIVE_AUTHENTICATION_ERROR:
            return {
                ...state,
                isFailed: true,
                isLoading: false,
                error: action.error
            };

        case CLEAR_AUTHENTICATION_DATA:
            return {
                payload: null,
                isFailed: false,
                isLoading: false,
                error: null
            };

        default: return state;
    };
};

export default loginReducer;