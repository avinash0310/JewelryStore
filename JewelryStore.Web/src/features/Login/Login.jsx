import React, { Component } from 'react';
import { connect } from "react-redux";
import { requestLogin, clearLoginData } from './actions';
import {
    INVALID_CREDENTIAL_VALIDATION_MSG, INVALID_USER_NAME_VALIDATION_MSG, DEFAULT_PAGE_PATH,
    INVALID_PASSWORD_VALIDATION_MSG, TOKEN, JEWELRY_STORE_PATH, MIN_LENGTH, MAX_LENGTH, FAILED_MESSAGE
} from '../../app/common/constant';

export class Login extends Component {
    constructor(props) {
        super(props);
        this.state = {
            userName: "",
            password: "",
            isDisabled: false
        };
        this.handleValidation = this.handleValidation.bind(this);
        this.handleSubmit = this.handleSubmit.bind(this);
        this.handleChange = this.handleChange.bind(this);
    };


    componentDidUpdate(prevProps) {
        if (this.props.loginInfo.payload && prevProps.loginInfo.payload !== this.props.loginInfo.payload) {
            if (this.props.loginInfo.payload.hasOwnProperty(TOKEN)) {
                this.props.history.push(JEWELRY_STORE_PATH);
            }
            else {
                this.setState({
                    userName: "",
                    password: "",
                    isDisabled: false
                }, () => alert(INVALID_CREDENTIAL_VALIDATION_MSG));
            };
        };

        if (this.props.loginInfo !== prevProps.loginInfo && this.props.loginInfo.isFailed) {
            this.setState({
                isDisabled: false
            }, () => alert(FAILED_MESSAGE));
        };
    };

    handleValidation = () => {
        let userNameLength = this.state.userName.trim().length;
        let passwordLength = this.state.password.trim().length;
        if (userNameLength === 0 || userNameLength < MIN_LENGTH || userNameLength > MAX_LENGTH) {
            alert(INVALID_USER_NAME_VALIDATION_MSG);
            return false;
        }
        else if (passwordLength === 0 || passwordLength < MIN_LENGTH || passwordLength > MAX_LENGTH) {
            alert(INVALID_PASSWORD_VALIDATION_MSG);
            return false;
        };

        return true;
    };

    handleSubmit = (event) => {
        event.preventDefault();
        if (this.handleValidation()) {
            var loginModel = {
                userName: this.state.userName,
                password: this.state.password
            };
            this.setState({ isDisabled: true })
            this.props.requestLogin(loginModel);
        };
    };

    handleCancel = (event) => {
        event.preventDefault();
        this.setState({
            userName: "",
            password: "",
            isDisabled: false
        }, () => {
            this.props.clearLoginData();
            this.props.history.push(DEFAULT_PAGE_PATH);
        });
    };

    handleChange = (event) => {
        this.setState({
            [event.target.name]: event.target.value
        });
    };

    render() {
        return (
            <div className="login-page">
                <div className="login-control">
                    <label>Username : </label>
                    <input type="text" name="userName" placeholder="Username" value={this.state.userName} onChange={this.handleChange} />
                </div>
                <div className="login-control">
                    <label>Password : </label>
                    <input type="password" name="password" placeholder="Password" value={this.state.password} onChange={this.handleChange} />
                </div>
                <div className="login-control-btn">
                    <button disabled={this.state.isDisabled} onClick={this.handleSubmit}>{this.state.isDisabled ? 'Validating Credential...' : 'Login'}</button>
                    <button onClick={this.handleCancel}>Cancel</button>
                </div>
            </div>
        )
    };
};


const mapStateToProps = state => {
    return {
        loginInfo: state.loginDetails,
    };
};

const mapDispatchToProps = dispatch => {
    return {
        requestLogin: (loginModel) => dispatch(requestLogin(loginModel)),
        clearLoginData: () => dispatch(clearLoginData())
    };
};

export default connect(mapStateToProps, mapDispatchToProps)(Login);