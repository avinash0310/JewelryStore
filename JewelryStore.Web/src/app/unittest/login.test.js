import React from 'react';
import { shallow } from 'enzyme';
import { Login } from '../../features/Login/Login.jsx';
import { INVALID_PASSWORD_VALIDATION_MSG, INVALID_USER_NAME_VALIDATION_MSG } from "../common/constant";

describe('Login', () => {
    const props = {
        dispatch: jest.fn(),
        loginInfo: {
            payload: null,
            isFailed: false,
            isLoading: false,
            error: null
        }
    };

    it('username check', () => {
        // when
        let wrapper = shallow(<Login {...props} />);
        wrapper.find('input[type="text"]').simulate('change', { target: { name: 'userName', value: 'test0001' } });

        // then
        expect(wrapper.state('userName')).toEqual('test0001');
    });

    it('password check', () => {
        // when
        let wrapper = shallow(<Login {...props} />);
        wrapper.find('input[type="password"]').simulate('change', { target: { name: 'password', value: 'test0001' } });

        // then
        expect(wrapper.state('password')).toEqual('test0001');
    });

    it('login check with wrong username', () => {
        // when
        window.alert = jest.fn();
        let wrapper = shallow(<Login {...props} />);
        wrapper.find('input[type="text"]').simulate('change', { target: { name: 'userName', value: 'test0014234234234' } });
        wrapper.find('input[type="password"]').simulate('change', { target: { name: 'password', value: 'test001' } });
        wrapper.find('button').first().simulate('click', {
            preventDefault: () => {
            }
        });

        // then
        expect(window.alert).toHaveBeenCalledWith(INVALID_USER_NAME_VALIDATION_MSG);
        expect(wrapper.state('isDisabled')).toBe(false);
    });

    it('login check with wrong password', () => {
        // when
        let wrapper = shallow(<Login {...props} />);
        window.alert = jest.fn();
        wrapper.find('input[type="text"]').simulate('change', { target: { name: 'userName', value: 'test001' } });
        wrapper.find('input[type="password"]').simulate('change', { target: { name: 'password', value: 'test0014234234234' } });
        wrapper.find('button').first().simulate('click', {
            preventDefault: () => {
            }
        });

        // then
        expect(window.alert).toHaveBeenCalledWith(INVALID_PASSWORD_VALIDATION_MSG);
        expect(wrapper.state('isDisabled')).toBe(false);
    });
});