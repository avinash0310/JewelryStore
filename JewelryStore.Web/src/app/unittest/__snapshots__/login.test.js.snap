// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login should render initial layout 1`] = `
Array [
  <Context.Provider
    value={
      Object {
        "store": Object {
          "@@observable": [Function],
          "dispatch": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "subscription": Subscription {
          "handleChangeWrapper": [Function],
          "listeners": Object {
            "notify": [Function],
          },
          "onStateChange": [Function],
          "parentSub": undefined,
          "store": Object {
            "@@observable": [Function],
            "dispatch": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
          },
          "unsubscribe": null,
        },
      }
    }
  >
    <Memo(Connect(Login)) />
  </Context.Provider>,
]
`;

exports[`MyComponent should render initial layout 1`] = `
Array [
  <Context.Provider
    value={
      Object {
        "store": Object {
          "@@observable": [Function],
          "dispatch": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "subscription": Subscription {
          "handleChangeWrapper": [Function],
          "listeners": Object {
            "notify": [Function],
          },
          "onStateChange": [Function],
          "parentSub": undefined,
          "store": Object {
            "@@observable": [Function],
            "dispatch": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
          },
          "unsubscribe": null,
        },
      }
    }
  >
    <Memo(Connect(Login)) />
  </Context.Provider>,
]
`;
